src/Init/Data/List/Lemmas.lean

```diff
@@ -1202,6 +1343,16 @@ theorem eq_replicate {a : α} {n} {l : List α} :
   ⟨fun h => h ▸ ⟨length_replicate .., fun _ => eq_of_mem_replicate⟩,
    fun ⟨e, al⟩ => e ▸ eq_replicate_of_mem al⟩
 
+theorem map_eq_replicate_iff {l : List α} {f : α → β} {b : β} :
+    l.map f = replicate l.length b ↔ ∀ x ∈ l, f x = b := by
+  simp [eq_replicate]
+
+@[simp] theorem map_const (l : List α) (b : β) : map (Function.const α b) l = replicate l.length b :=
+  map_eq_replicate_iff.mpr fun _ _ => rfl
+
+@[simp] theorem map_const' (l : List α) (b : β) : map (fun _ => b) l = replicate l.length b :=
```

Comment:
Looks very useful, but I assume this will now be tried for every and each `List.map` application out there, due to the discrtree structure, will it? Should we worry about the cost?

Reply:
Yes, this is terrible, we can't do that! :-)

Reply:
But then we also cannot have `map (fun x => x) l = l` either, can we, which would be sad.

Maybe we need to add support for lambdas to the discr tree…